forc-plugins/forc-crypto/src/keys/vanity.rs

```diff
@@ -0,0 +1,551 @@
+        }
+        (_, Some(left), _, Some(right)) => {
+            let matcher = RegexMatcher { left, right };
+            find_vanity_address_with_timeout(matcher, mnemonic, timeout)
+        }
+        (_, Some(left), Some(right), _) => {
+            let matcher = LeftRegexRightExactMatcher { left, right };
+            find_vanity_address_with_timeout(matcher, mnemonic, timeout)
+        }
+        (Some(left), None, None, None) => {
+            let matcher = LeftHexMatcher { left };
+            find_vanity_address_with_timeout(matcher, mnemonic, timeout)
+        }
+        (None, None, Some(right), None) => {
+            let matcher = RightHexMatcher { right };
+            find_vanity_address_with_timeout(matcher, mnemonic, timeout)
+        }
+        (None, Some(re), None, None) => {
+            let matcher = SingleRegexMatcher { re };
+            find_vanity_address_with_timeout(matcher, mnemonic, timeout)
+        }
+        (None, None, None, Some(re)) => {
+            let matcher = SingleRegexMatcher { re };
+            find_vanity_address_with_timeout(matcher, mnemonic, timeout)
+        }
+        _ => {
+            return Err(VanityAddressError::InvalidPattern(
+                "Invalid pattern combination".to_string(),
+            )
+            .into())
+        }
+    };
+
+    let (address, secret_key, mnemonic) = result?;
+
+    let duration = start_time.elapsed();
+    println!(
+        "Successfully found vanity address in {:.3} seconds.\n",
+        duration.as_secs_f64()
+    );
+
+    let result = if let Some(mnemonic) = mnemonic {
+        json!({
+            "Address": address.to_string(),
+            "PrivateKey": hex::encode(secret_key.as_ref()),
+            "Mnemonic": mnemonic,
+        })
+    } else {
+        json!({
+            "Address": address.to_string(),
+            "PrivateKey": hex::encode(secret_key.as_ref()),
+        })
+    };
+
+    if let Some(path) = save_path {
+        fs::write(path, serde_json::to_string_pretty(&result)?)?;
+    }
+
+    Ok(result)
+}
+
+pub fn find_vanity_address_with_timeout<T: VanityMatcher>(
+    matcher: T,
+    use_mnemonic: bool,
+    timeout_secs: Option<u64>,
+) -> anyhow::Result<(Address, SecretKey, Option<String>)> {
+    let generate_wallet = move || {
+        let breakpoint = if use_mnemonic { 1_000 } else { 100_000 };
+        let start = Instant::now();
+        let attempts = std::sync::atomic::AtomicUsize::new(0); // atomic for parallel iteration
+        wallet_generator(use_mnemonic)
+            .find_any(|result| {
+                let current = attempts.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
+                if current != 0 && current % breakpoint == 0 {
+                    let elapsed = start.elapsed().as_secs_f64();
+                    let rate = current as f64 / elapsed;
+                    println!(
+                        "└─ tried {} addresses ({:.2} addresses/sec)...",
```

Comment:
If you want to make it look extra cool, you could use [indicatif](https://github.com/console-rs/indicatif/tree/main) and make this statement update the address count and rate update on the same line, similar to [this example](https://github.com/console-rs/indicatif/blob/main/examples/yarnish.rs#L84)

Reply:
Oh nice, indicatif looks really cool!

Reply:
had a look; cant think of a great UI; also this dependency is not currently in the project.
Maybe a future enhancement.